[JSC] AsyncGenerator should inline first queue item

Constellation opened the pull request:
#### b17f39692c3d55ec4ca4c49a08cd78129abc701f
<pre>
[JSC] AsyncGenerator should inline first queue item
<a href="https://bugs.webkit.org/show_bug.cgi?id=304783">https://bugs.webkit.org/show_bug.cgi?id=304783</a>
<a href="https://rdar.apple.com/167333858">rdar://167333858</a>

Reviewed by Sosuke Suzuki.

In common case, only one queue item is enqueued into JSAsyncGenerator.
This patch inlines this queue item into JSAsyncGenerator itself. We move
enqueue and dequeue of JSAsyncGenerator completely into C++
(JSAsyncGenerator::enqueue / JSAsyncGenerator::dequeue), and inside it,
we will inline the first queue item. Furthermore we hide this queueing
mechanism from JS code so that we can use JSPromiseReaction instead,
which is much more efficient.

Also we clean up AsyncGeneratorPrototype.js and C++ code, folding things
for faster performance and more importantly, using explicit loop for
repeated invocation (in particular for C++ to avoid recursion).

* Source/JavaScriptCore/builtins/AsyncGeneratorPrototype.js:
(linkTimeConstant.asyncGeneratorResumeNext):
(next):
(return):
(throw):
(linkTimeConstant.asyncGeneratorQueueIsEmpty): Deleted.
(linkTimeConstant.asyncGeneratorQueueEnqueue): Deleted.
(linkTimeConstant.asyncGeneratorQueueDequeue): Deleted.
(linkTimeConstant.isExecutionState): Deleted.
(linkTimeConstant.isSuspendYieldState): Deleted.
(linkTimeConstant.asyncGeneratorReject):
(linkTimeConstant.asyncGeneratorResolve):
(linkTimeConstant.asyncGeneratorEnqueue):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.cpp:
(JSC::BytecodeIntrinsicRegistry::BytecodeIntrinsicRegistry):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::asyncGeneratorInternalFieldIndex):
* Source/JavaScriptCore/runtime/JSAsyncGenerator.cpp:
(JSC::JSAsyncGenerator::enqueue):
* Source/JavaScriptCore/runtime/JSAsyncGenerator.h:
* Source/JavaScriptCore/runtime/JSCast.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::asyncGeneratorReject):
(JSC::asyncGeneratorResolve):
(JSC::doAsyncGeneratorBodyCall):
(JSC::asyncGeneratorResumeNext):
(JSC::asyncGeneratorYieldAwaited):
(JSC::asyncGeneratorBodyCallNormal):
(JSC::asyncGeneratorBodyCallReturn):
(JSC::asyncGeneratorResumeNextReturn):
(JSC::asyncGeneratorQueueIsEmpty): Deleted.
(JSC::asyncGeneratorQueueDequeue): Deleted.
(JSC::asyncGeneratorYield): Deleted.

Canonical link: <a href="https://commits.webkit.org/305016@main">https://commits.webkit.org/305016@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc60cd57e7c8ff62fdef773f8eae24636136e820

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137189 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9549 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48476 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144936 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90162 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/f53c5fb5-c698-4b45-8686-2232b4eba742) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10252 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9676 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104916 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/1b90b157-1866-4a3e-af74-178f539bccca) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140134 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7555 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122942 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85755 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7192 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4903 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5525 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/129150 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116547 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41106 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147694 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/135681 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9232 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41668 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113274 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9250 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7767 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113604 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7113 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119198 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63661 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21138 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9280 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37252 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/168457 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/9006 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72846 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43970 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9221 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/9073 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
```diff
@@ -844,6 +847,53 @@ JSC_DEFINE_HOST_FUNCTION(performPromiseThen, (JSGlobalObject* globalObject, Call
+    JSPromise* promise = jsCast<JSPromise*>(callFrame->uncheckedArgument(3));
+
+    if (!generator) [[unlikely]] {
+        promise->reject(vm, globalObject, createTypeError(globalObject, "|this| should be an async generator"_s));
```
Don't we need to add exception check here?

Constellation replied:
Ditto

sosukesuzuki commented on Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
```diff
@@ -844,6 +847,53 @@ JSC_DEFINE_HOST_FUNCTION(performPromiseThen, (JSGlobalObject* globalObject, Call
+
+    auto [value, resumeMode, promise] = generator->dequeue(vm);
+
+    promise->resolve(globalObject, resolution);
```
ditto

Constellation replied:
It is fine since this is the last expression and we are not having throwScope in this function.

sosukesuzuki commented on Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
```diff
@@ -844,6 +847,53 @@ JSC_DEFINE_HOST_FUNCTION(performPromiseThen, (JSGlobalObject* globalObject, Call
+
+    auto [value, resumeMode, promise] = generator->dequeue(vm);
+
+    promise->reject(vm, globalObject, error);
```
ditto

Constellation replied:
Not necessary. It never throws.

sosukesuzuki reviewed (approved):
r=me

Constellation reviewed (commented):
Replied

Constellation committed:
[JSC] AsyncGenerator should inline first queue item
https://bugs.webkit.org/show_bug.cgi?id=304783
rdar://167333858

Reviewed by Sosuke Suzuki.

In common case, only one queue item is enqueued into JSAsyncGenerator.
This patch inlines this queue item into JSAsyncGenerator itself. We move
enqueue and dequeue of JSAsyncGenerator completely into C++
(JSAsyncGenerator::enqueue / JSAsyncGenerator::dequeue), and inside it,
we will inline the first queue item. Furthermore we hide this queueing
mechanism from JS code so that we can use JSPromiseReaction instead,
which is much more efficient.

Also we clean up AsyncGeneratorPrototype.js and C++ code, folding things
for faster performance and more importantly, using explicit loop for
repeated invocation (in particular for C++ to avoid recursion).

* Source/JavaScriptCore/builtins/AsyncGeneratorPrototype.js:
(linkTimeConstant.asyncGeneratorResumeNext):
(next):
(return):
(throw):
(linkTimeConstant.asyncGeneratorQueueIsEmpty): Deleted.
(linkTimeConstant.asyncGeneratorQueueEnqueue): Deleted.
(linkTimeConstant.asyncGeneratorQueueDequeue): Deleted.
(linkTimeConstant.isExecutionState): Deleted.
(linkTimeConstant.isSuspendYieldState): Deleted.
(linkTimeConstant.asyncGeneratorReject):
(linkTimeConstant.asyncGeneratorResolve):
(linkTimeConstant.asyncGeneratorEnqueue):
* Source/JavaScriptCore/builtins/BuiltinNames.h:
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.cpp:
(JSC::BytecodeIntrinsicRegistry::BytecodeIntrinsicRegistry):
* Source/JavaScriptCore/bytecode/BytecodeIntrinsicRegistry.h:
* Source/JavaScriptCore/bytecode/LinkTimeConstant.h:
* Source/JavaScriptCore/bytecompiler/NodesCodegen.cpp:
(JSC::asyncGeneratorInternalFieldIndex):
* Source/JavaScriptCore/runtime/JSAsyncGenerator.cpp:
(JSC::JSAsyncGenerator::enqueue):
* Source/JavaScriptCore/runtime/JSAsyncGenerator.h:
* Source/JavaScriptCore/runtime/JSCast.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
(JSC::JSGlobalObject::init):
* Source/JavaScriptCore/runtime/JSMicrotask.cpp:
(JSC::asyncGeneratorReject):
(JSC::asyncGeneratorResolve):
(JSC::doAsyncGeneratorBodyCall):
(JSC::asyncGeneratorResumeNext):
(JSC::asyncGeneratorYieldAwaited):
(JSC::asyncGeneratorBodyCallNormal):
(JSC::asyncGeneratorBodyCallReturn):
(JSC::asyncGeneratorResumeNextReturn):
(JSC::asyncGeneratorQueueIsEmpty): Deleted.
(JSC::asyncGeneratorQueueDequeue): Deleted.
(JSC::asyncGeneratorYield): Deleted.

Canonical link: https://commits.webkit.org/305016@main
